sdks/python/apache_beam/runners/interactive/caching/filebasedcache.py

```diff
@@ -0,0 +1,276 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import urllib
+import warnings
+
+from apache_beam import coders
+from apache_beam.io import avroio
+from apache_beam.io import parquetio
+from apache_beam.io import textio
+from apache_beam.io import tfrecordio
+from apache_beam.io.filesystems import FileSystems
+from apache_beam.runners.interactive.caching import PCollectionCache
+from apache_beam.runners.interactive.caching.datatype_inference import infer_element_type
+from apache_beam.transforms import PTransform
+
+try:
+  from weakref import finalize
+except ImportError:
+  from backports.weakref import finalize
+
+try:  # Python 3
+  unquote_to_bytes = urllib.parse.unquote_to_bytes
+  quote = urllib.parse.quote
+except AttributeError:  # Python 2
+  # pylint: disable=deprecated-urllib-function
+  unquote_to_bytes = urllib.unquote
+  quote = urllib.quote
+
+__all__ = [
+    "FileBasedCache",
+    "TextBasedCache",
+    "SafeTextBasedCache",
+    "TFRecordBasedCache",
+    "ParquetBasedCache",
+    "AvroBasedCache",
+    "SafeFastPrimitivesCoder",
+]
+
+
+class FileBasedCache(PCollectionCache):
+
+  def __init__(self, location, mode="error", persist=False, **writer_kwargs):
+    self.location = location
+    self._writer_kwargs = writer_kwargs
+    self.coder = writer_kwargs.pop("coder", None)
+    self._coder_was_provided = self.coder is not None
+    self._num_writes = 0
+    self._persist = persist
+    self._finalizer = (
+        finalize(self, lambda: None) if self._persist else finalize(
+            self, lambda pattern: FileSystems.delete(
+                list(glob_files(pattern + "**"))), self.file_pattern))
+
+    # TODO(ostrokach): Implement append mode.
+    if mode not in ['error', 'overwrite']:
+      raise ValueError("'mode' must be set to one of: ['error', 'overwrite'].")
+    exitsting_files = list(glob_files(self.file_pattern))
+    if mode == "error" and exitsting_files:
+      raise IOError("The following cache files already exist: {}.".format(
+          exitsting_files))
+    if mode == "overwrite":
+      self.truncate()
+
+    root, _ = FileSystems.split(self._file_path_prefix)
+    try:
+      FileSystems.mkdirs(root)
+    except IOError:
+      pass
+    # It is possible to read from am empty stream, so it should also be possible
+    # to read from an empty file.
+    FileSystems.create(self._file_path_prefix + ".empty").close()
+
+  @property
+  def persist(self):
+    return self._persist
+
+  @persist.setter
+  def persist(self, persist):
+    self._persist = persist
+    self._finalizer = (lambda: None) if self._persist else finalize(
```

Comment:
Yes, thank you, this was a serious mistake on my part! I fixed the persist logic and added a test for the intended behavior.

The idea behind `cache.persist = True` is to make it possible to persist cache between sessions. With Jupyter notebooks, it is often the case that the user has to restart the kernel, for example if they ran out of memory or if they want to reload a dependency. In order to facilitate this use-case, we may need a way to persist cache between sessions. This could be accomplished by setting `cache.persist = True` for every cache and saving the cache instances to a `shelve`.